pkg/dependency/parser/dotnet/core_deps/parse.go

```diff
@@ -55,20 +87,36 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]ftypes.Package, []ftypes.Dependenc
 					EndLine:   lib.EndLine,
 				},
 			},
+			// We're still not sure that we need to skip libraries built into .NETCore (or that we detect them correctly).
+			// So we mark these libraries as Dev to skip the scan by default, but keep the options for displaying these libraries.
```

Comment:
in general I think the same way.
thank you for detailing this :+1:

> In all cases - if it's bits don't make it to the final app it cannot be a source of a vulnerability in that app.

There are times when users want to see dependencies that are not used at runtime (we encountered this in nodejs).
That's why we've added options to allow users to see all dependencies (we hide them by default). for example, this may be needed not for scanning vulnerabilities, but for generating a SBOM

Reply:
> The package might be referenced with ExcludeAssets="Runtime"

Do we need to check this case?
Can you write more info about this?

[docs](https://github.com/dotnet/sdk/blob/529132850841a6bcfce96799262ce688e3851875/documentation/specs/runtime-configuration-file.md#targets-section-depsjson) don't have info about `ExcludeAssets` field.

Reply:
> in general I think the same way.
> 
> thank you for detailing this :+1:
> 
> 
> 
> > In all cases - if it's bits don't make it to the final app it cannot be a source of a vulnerability in that app.
> 
> 
> 
> There are times when users want to see dependencies that are not used at runtime (we encountered this in nodejs).
> 
> That's why we've added options to allow users to see all dependencies (we hide them by default). for example, this may be needed not for scanning vulnerabilities, but for generating a SBOM
> 
> 

I would add more contexts. As interest in supply chain security grows, there is a demand to understand vulnerabilities in packages that are used for CI/CD and other purposes, even if they are not included in production applications and are used only for development.

If the library is not used at all, even for development purposes, there is no need to include it.

Reply:
@ericstj Since we're unfamiliar with .NET Core, we need your input. Given the above context, do you think we don't have to save these libraries?

Reply:
> [docs](https://github.com/dotnet/sdk/blob/529132850841a6bcfce96799262ce688e3851875/documentation/specs/runtime-configuration-file.md#targets-section-depsjson) don't have info about ExcludeAssets field.

You can find the docs on ExcludeAssets [here](https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files).

> do you think we don't have to save these libraries?

They may be interesting for informational purposes, but for the actual CVE which apply to runtime libraries (the vast majority of CVEs) they are not interesting.  

There could be a case where a package only provides development time assets.  For instance - a build target that inserts code into the binary.  That package might appear without any runtime assets, however such a package is not the norm.  Such a package could also remove all traces from the deps.json if it wanted to, so trying to use that for detecting it isn't a legitimate use case.  The deps.json isn't a manifest of the things used when building the project - it's a manifest to be used by the runtime to determine what it can load.

Reply:
> You can find the docs on ExcludeAssets [here](https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files).

.NET uses so many files that it can get confusing... 
IIUC `PackageReference` field is related with `packages.config` and `*Packages.props` files (`xml` format).
Can this field be used for *.deps.json files?

> They may be interesting for informational purposes, but for the actual CVE which apply to runtime libraries (the vast majority of CVEs) they are not interesting.

If there are cases when non-runtime dependencies might be needed - i think we need to hide them (instead of remove them). As [@knqyf263 correctly said](https://github.com/aquasecurity/trivy/pull/7039#discussion_r1658126726) DevOps engineers may need to check these packages, for example in pipelines.

Reply:
In other ecosystems, for example, it is common to include test assertion libraries as development libraries. And while these libraries are not used at runtime, their vulnerability is important.

>For instance - a build target that inserts code into the binary. That package might appear without any runtime assets, however such a package is not the norm. .

This example you gave is certainly not legitimate, but are libraries used for development purposes in general use such as the one I gave not included in deps.json?

Reply:
`deps.json` isn't the file to see all dependencies that might be used at development time.  It's used for loading binaries at runtime. 

At development time projects use `PackageReference` (direct references) in the project, which gets computed to the full package graph in `project.assets.json`.  The old format `packages.config` had a flattened graph in the single file and was really just a log for the package modifications that were made to the project file separately.  These two files are better for identifying development dependencies - and are used by component governance for that purpose.  They are present on the build machine / source control.  There's also the new SBOM infrastructure which is better suited as a catch all.

> This example you gave is certainly not legitimate, but are libraries used for development purposes in general use such as the one I gave not included in deps.json?

They may or may not be included, it depends on the type of package and how its referenced.  `deps.json` is only used for telling the host what to load at runtime.  When a library is present you can be certain it's used - so if that library/package is vulnerable it should be flagged.  If the package is listed in the deps file, you can be certain it was at least referenced - but it may have been eliminated by the build and not loadable at runtime - so it should not be flagged.  You can't assume that a package being absent means it wasn't used at dev time since the deps.json doesn't list all development time dependencies.

Reply:
hm.. It looks like that we can remove non-runtime dependencies.
Also we need to write in docs that Trivy only detects runtime deps from `*.deps.json`

Reply:
Thanks for your explanation. It might be similar to Go binary scanning. It doesn't have any development modules, while go.mod includes it.

@DmitriyLewen We can just delete non-runtime dependencies and improve it later if we see any feedback.

Reply:
Done :+1: 
@knqyf263 @ericstj can you review this PR?